### ISSUE TITLE
Migrate: rollback down functions in reverse order from up functions

### ISSUE THREAD

**Issue by dealercontrolrachel** [tgriesser/knex#2832]
# Environment

Knex version: 0.15.2
Database + version: Mysql 8.0 Ubuntu docker container
OS: macOs High Sierra 10.13.6

# Feature discussion / request

There may be a solution to this and if so, I apologize but I cannot figure it out. 

1.
I have lots of migrations files that all create and drop tables. There are some foreign keys that reference tables in different files. 

ex:
 - base_migration: up creates applications table, down drops it
 - users_migration: up creates users table and user_applications_link (foreign keys to users and applications), down drops the link and then the users. 

so the problem is that when I rollback, knex runs the down functions in the same order as the ups and tries to drop the applications table before that link table that has a foreign key reference to it. 

2, 3.

I would like the ability to do some sort of `knex migtate:rollback -reverse` or something like that so that the users file down function could run first and then the base down function. Or just have a setting somewhere that specifies that I want the down functions to run in reverse order.

Currently I have all of the drop table commands in the base file and the other files just return a promise but I don't like this very much. I realize I could have all the linked tables in the same file but this will not work in all situations like I have a phone_type_list which like 6 tables in all different modules have foreign key references to. 

Please let me know if anyone has any suggestions.

**Comment by elhigu** [comment]
> so the problem is that when I rollback, knex runs the down functions in the same order as the ups and tries to drop the applications table before that link table that has a foreign key reference to it.

That sounds like a bug if it indeed does that. Should test it out to make sure (I never use down -scripts myself)

**Comment by dealercontrolrachel** [comment]
Okay I don't know what happened but it seems to be working fine now. I swear it was going in reverse. I put console.logs in the down functions. I was also using knex-migrate so maybe that was doing something weird?

**Comment by danielscotiabank** [comment]
Hi there! I'm bringing this back to life because I've been able to reproduce the issue with knex 0.16.5. Some context: my organization has a quality gate that requires enough test coverage and my build was failing because the migration files were being picked up. While there are ways around it I decided to entertain the possibility of testing my migrations at build time with an in-memory database.

The idea is to start with a blank database, apply all migrations and then roll them all back. Here is the code:

`// src/migrations/common/201905061100_create_my_table.js

exports.up = knex => {
	return knex.schema.createTable('my_table', table => {
		table.string('key').notNullable().primary();
		table.text('value').notNullable();
	});
};

exports.down = knex => {
	return knex.schema.dropTable('my_table');
};

// src/migrations/dev/201905061200_insert_my_table.js

exports.up = knex => {
	return knex.table('my_table').insert({
		key: 'my_key',
		value: 'my_value'
	});
};

exports.down = knex => {
	return knex.table('my_table').where({
		key: 'my_key'
	}).del();
};

// migration_test.js

... set up knex with sqlite ...

describe('migrations', () => {

	const config = {
		directory: [
			__dirname + "/../src/migrations/common",
			__dirname + "/../src/migrations/dev"
		]
	}
		
	it('execute successfully', done => {
		knex.migrate
				.latest(config)
				.then(() => migrator
						.rollback(config, true)
						.then(() => done()))
				.catch(error => done(error))
	})

})`

Here are the logs:

`$ NODE_ENV=dev DEBUG=knex:query npm test

...

  migrations for
2019-05-09T17:43:53.591Z knex:query select * from sqlite_master where type = 'table' and name = ? undefined
2019-05-09T17:43:53.604Z knex:query create table `knex_migrations` (`id` integer not null primary key autoincrement, `name` varchar(255), `batch` integer, `migration_time` datetime) undefined
2019-05-09T17:43:53.606Z knex:query select * from sqlite_master where type = 'table' and name = ? undefined
2019-05-09T17:43:53.607Z knex:query create table `knex_migrations_lock` (`index` integer not null primary key autoincrement, `is_locked` integer) undefined
2019-05-09T17:43:53.612Z knex:query select * from `knex_migrations_lock` undefined
2019-05-09T17:43:53.613Z knex:query insert into `knex_migrations_lock` (`is_locked`) values (?) undefined
2019-05-09T17:43:53.614Z knex:query select `name` from `knex_migrations` order by `id` asc undefined
2019-05-09T17:43:53.714Z knex:query BEGIN; trx2
2019-05-09T17:43:53.715Z knex:query select * from `knex_migrations_lock` trx2
2019-05-09T17:43:53.717Z knex:query update `knex_migrations_lock` set `is_locked` = ? trx2
2019-05-09T17:43:53.717Z knex:query select * from sqlite_master where type = 'table' and name = ? trx2
2019-05-09T17:43:53.718Z knex:query select * from sqlite_master where type = 'table' and name = ? trx2
2019-05-09T17:43:53.718Z knex:query select * from `knex_migrations_lock` trx2
2019-05-09T17:43:53.719Z knex:query select `name` from `knex_migrations` order by `id` asc trx2
2019-05-09T17:43:53.719Z knex:query select max(`batch`) as `max_batch` from `knex_migrations` trx2
**2019-05-09T17:43:53.721Z knex:query create table `my_table` (`key` varchar(255) not null, `value` text not null, primary key (`key`)) trx2**
2019-05-09T17:43:53.722Z knex:query insert into `knex_migrations` (`batch`, `migration_time`, `name`) values (?, ?, ?) trx2
**2019-05-09T17:43:53.723Z knex:query insert into `my_table` (`value`, `key`) values (?, ?) trx2**
2019-05-09T17:43:53.723Z knex:query insert into `knex_migrations` (`batch`, `migration_time`, `name`) values (?, ?, ?) trx2
2019-05-09T17:43:53.723Z knex:query update `knex_migrations_lock` set `is_locked` = ? trx2
2019-05-09T17:43:53.724Z knex:query COMMIT; trx2
2019-05-09T17:43:53.725Z knex:query select * from sqlite_master where type = 'table' and name = ? trx2
2019-05-09T17:43:53.725Z knex:query select * from sqlite_master where type = 'table' and name = ? trx2
2019-05-09T17:43:53.726Z knex:query select * from `knex_migrations_lock` trx2
2019-05-09T17:43:53.726Z knex:query select `name` from `knex_migrations` order by `id` asc trx2
2019-05-09T17:43:53.727Z knex:query BEGIN; trx3
2019-05-09T17:43:53.728Z knex:query select * from `knex_migrations_lock` trx3
2019-05-09T17:43:53.728Z knex:query update `knex_migrations_lock` set `is_locked` = ? trx3
2019-05-09T17:43:53.729Z knex:query COMMIT; trx3
2019-05-09T17:43:53.729Z knex:query select max(`batch`) as `max_batch` from `knex_migrations` trx3
2019-05-09T17:43:53.730Z knex:query BEGIN; trx4
**2019-05-09T17:43:53.731Z knex:query drop table `my_table` trx4**
2019-05-09T17:43:53.731Z knex:query COMMIT; trx4
2019-05-09T17:43:53.732Z knex:query delete from `knex_migrations` where `name` = ? trx4
2019-05-09T17:43:53.733Z knex:query BEGIN; trx5
**2019-05-09T17:43:53.734Z knex:query delete from `my_table` where `key` = ? trx5**
2019-05-09T17:43:53.735Z knex:query ROLLBACK trx5
migration file "20190506120000_insert_my_table.js" failed
migration failed with error: delete from `my_table` where `key` = 'my_key' - SQLITE_ERROR: no such table: my_table
2019-05-09T17:43:53.736Z knex:query update `knex_migrations_lock` set `is_locked` = ? trx5
    1) dev execute successfully


  0 passing (176ms)
  1 failing

  1) migrations for
       dev execute successfully:
     delete from `my_table` where `key` = 'my_key' - SQLITE_ERROR: no such table: my_table
  Error: SQLITE_ERROR: no such table: my_table

npm ERR! Test failed.  See above for more details.
`

In the actual production code I have transferred the insert to a seed file. However, if I wanted to, say, create a column and drop it in the rollback process, wouldn't it behave the same way?

Thanks!
